File: Week4/BinarySearchProblem/SearchInsertPosition.java
class SearchInsertPosition{

/*
 * Given a sorted array of distinct integers and a target value, return the index if the target is 
 * found. If not, return the index where it would be if it were inserted in order.
 * 
 * https://leetcode.com/problems/search-insert-position/
 * T(c) -> O(logn), S(c) -> O(logn)
 */
    static int searchInsert(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;
        while(low <= high)
        {
            int mid = low + (high-low)/2;
            // element found
            if(nums[mid] == target)
            {
                System.out.println("Element Found");
                return mid;
            }
            // go to left part
            else if(nums[mid] > target)
            {
                high = mid-1;
            }
            // go to right part
            else{
                low = mid+1;
            }
        }
        // return insert position if element not found
        return low;
    }

    public static void main(String[] args) {
        int arr[] = {1,3,5,6};
        int target = 2;
        System.out.println("Index = " + searchInsert(arr, target));
    }

}

File: Week4/BinarySearchProblem/python/SearchInsertPosition.py
"""
Given a sorted array of distinct integers and a target value, return the index if the target is found. If not, return 
the index where it would be if it were inserted in order.

Input: nums = [1,3,5,6], target = 2
Output: 1

https://leetcode.com/problems/search-insert-position/description/
"""

from typing import List

def searchInsert(nums: List[int], target: int) -> int:
    """ T(c) -> O(logn), S(c) -> O(1) """
    low = 0
    high = len(nums)-1
    ans = len(nums)
    # using lower bound method to find which index can contain the search element 
    while low <= high:
        mid = (low+high) // 2
        # first occurance (smallest index num >= target)
        if nums[mid] >= target:
            ans = mid
            # go to left side
            high = mid-1
        else:
            # go to right side
            low = mid+1

    return ans


nums = [1,3,5,6]
target = 2
print(nums)
print(target , " -> ", searchInsert(nums, target))
